--- graphite-web-0.9.8/webapp/graphite/storage.py	2011-04-01 20:48:43.000000000 +0100
+++ graphite-bzr/webapp/graphite/storage.py	2011-04-15 18:19:44.000000000 +0100
@@ -309,12 +309,13 @@
 
 class RRDFile(Branch):
   def getDataSources(self):
-    try:
-      info = rrdtool.info(self.fs_path)
-      return [RRDDataSource(self, source) for source in info['ds']]
-    except:
-      raise
-      return []
+    info = rrdtool.info(self.fs_path)
+    if 'ds' in info:
+      return [RRDDataSource(self, datasource_name) for datasource_name in info['ds']]
+    else:
+      ds_keys = [ key for key in info if key.startswith('ds[') ]
+      datasources = set( key[3:].split(']')[0] for key in ds_keys )
+      return [ RRDDataSource(self, ds) for ds in datasources ]
 
 
 class RRDDataSource(Leaf):
